### ISSUE TITLE
[inductor][cpu] FP32/AMP models multiple/single thread static/dynamic shape default/CPP wrapper accuracy crash in 2024-04-14 nightly release

### ISSUE THREAD

**Issue by zxd1997066** [pytorch/pytorch#124286]
### 🐛 Describe the bug

<p>Below models FP32/AMP datatype multiple/single thread static/dynamic shape default/CPP wrapper accuracy crash in 2024-04-14 nightly release</p><table border="1" class="dataframe table">
  <thead>
    <tr style="text-align: right;">
      <th>suite</th>
      <th>name</th>
      <th>thread</th>
      <th>accuracy</th>
      <th>perf</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>torchbench</td>
      <td>dcgan</td>
      <td>multiple</td>
      <td>X</td>
      <td>√</td>
    </tr>
    <tr>
      <td>torchbench</td>
      <td>densenet121</td>
      <td>multiple</td>
      <td>X</td>
      <td>√</td>
    </tr>
    <tr>
      <td>torchbench</td>
      <td>mnasnet1_0</td>
      <td>multiple</td>
      <td>X</td>
      <td>√</td>
    </tr>
    <tr>
      <td>torchbench</td>
      <td>mobilenet_v2</td>
      <td>multiple</td>
      <td>X</td>
      <td>√</td>
    </tr>
    <tr>
      <td>torchbench</td>
      <td>resnet152</td>
      <td>multiple</td>
      <td>X</td>
      <td>√</td>
    </tr>
    <tr>
      <td>torchbench</td>
      <td>resnet18</td>
      <td>multiple</td>
      <td>X</td>
      <td>√</td>
    </tr>
    <tr>
      <td>torchbench</td>
      <td>resnet50</td>
      <td>multiple</td>
      <td>X</td>
      <td>√</td>
    </tr>
    <tr>
      <td>torchbench</td>
      <td>resnext50_32x4d</td>
      <td>multiple</td>
      <td>X</td>
      <td>√</td>
    </tr>
    <tr>
      <td>torchbench</td>
      <td>shufflenet_v2_x1_0</td>
      <td>multiple</td>
      <td>X</td>
      <td>√</td>
    </tr>
    <tr>
      <td>timm_models</td>
      <td>dla102</td>
      <td>multiple</td>
      <td>X</td>
      <td>√</td>
    </tr>
    <tr>
      <td>timm_models</td>
      <td>ghostnet_100</td>
      <td>multiple</td>
      <td>X</td>
      <td>√</td>
    </tr>
    <tr>
      <td>timm_models</td>
      <td>hrnet_w18</td>
      <td>multiple</td>
      <td>X</td>
      <td>√</td>
    </tr>
    <tr>
      <td>timm_models</td>
      <td>levit_128</td>
      <td>multiple</td>
      <td>X</td>
      <td>√</td>
    </tr>
    <tr>
      <td>timm_models</td>
      <td>pnasnet5large</td>
      <td>multiple</td>
      <td>X</td>
      <td>√</td>
    </tr>
    <tr>
      <td>timm_models</td>
      <td>res2net101_26w_4s</td>
      <td>multiple</td>
      <td>X</td>
      <td>√</td>
    </tr>
    <tr>
      <td>timm_models</td>
      <td>res2net50_14w_8s</td>
      <td>multiple</td>
      <td>X</td>
      <td>√</td>
    </tr>
    <tr>
      <td>timm_models</td>
      <td>res2next50</td>
      <td>multiple</td>
      <td>X</td>
      <td>√</td>
    </tr>
    <tr>
      <td>timm_models</td>
      <td>resnest101e</td>
      <td>multiple</td>
      <td>X</td>
      <td>√</td>
    </tr>
    <tr>
      <td>timm_models</td>
      <td>rexnet_100</td>
      <td>multiple</td>
      <td>X</td>
      <td>√</td>
    </tr>
    <tr>
      <td>timm_models</td>
      <td>selecsls42b</td>
      <td>multiple</td>
      <td>X</td>
      <td>√</td>
    </tr>
    <tr>
      <td>timm_models</td>
      <td>swsl_resnext101_32x16d</td>
      <td>multiple</td>
      <td>X</td>
      <td>√</td>
    </tr>
    <tr>
      <td>timm_models</td>
      <td>visformer_small</td>
      <td>multiple</td>
      <td>X</td>
      <td>√</td>
    </tr>
    <tr>
      <td>timm_models</td>
      <td>volo_d1_224</td>
      <td>multiple</td>
      <td>X</td>
      <td>√</td>
    </tr>
    <tr>
      <td>timm_models</td>
      <td>xcit_large_24_p8_224</td>
      <td>multiple</td>
      <td>X</td>
      <td>√</td>
    </tr>
    <tr>
      <td>torchbench</td>
      <td>densenet121</td>
      <td>single</td>
      <td>X</td>
      <td>√</td>
    </tr>
    <tr>
      <td>torchbench</td>
      <td>mnasnet1_0</td>
      <td>single</td>
      <td>X</td>
      <td>√</td>
    </tr>
    <tr>
      <td>torchbench</td>
      <td>mobilenet_v2</td>
      <td>single</td>
      <td>X</td>
      <td>√</td>
    </tr>
    <tr>
      <td>torchbench</td>
      <td>resnet152</td>
      <td>single</td>
      <td>X</td>
      <td>√</td>
    </tr>
    <tr>
      <td>torchbench</td>
      <td>resnet18</td>
      <td>single</td>
      <td>X</td>
      <td>√</td>
    </tr>
    <tr>
      <td>torchbench</td>
      <td>resnet50</td>
      <td>single</td>
      <td>X</td>
      <td>√</td>
    </tr>
    <tr>
      <td>torchbench</td>
      <td>resnext50_32x4d</td>
      <td>single</td>
      <td>X</td>
      <td>√</td>
    </tr>
    <tr>
      <td>torchbench</td>
      <td>shufflenet_v2_x1_0</td>
      <td>single</td>
      <td>X</td>
      <td>√</td>
    </tr>
    <tr>
      <td>timm_models</td>
      <td>dla102</td>
      <td>single</td>
      <td>X</td>
      <td>√</td>
    </tr>
    <tr>
      <td>timm_models</td>
      <td>ghostnet_100</td>
      <td>single</td>
      <td>X</td>
      <td>√</td>
    </tr>
    <tr>
      <td>timm_models</td>
      <td>hrnet_w18</td>
      <td>single</td>
      <td>X</td>
      <td>√</td>
    </tr>
    <tr>
      <td>timm_models</td>
      <td>levit_128</td>
      <td>single</td>
      <td>X</td>
      <td>√</td>
    </tr>
    <tr>
      <td>timm_models</td>
      <td>pnasnet5large</td>
      <td>single</td>
      <td>X</td>
      <td>√</td>
    </tr>
    <tr>
      <td>timm_models</td>
      <td>res2net101_26w_4s</td>
      <td>single</td>
      <td>X</td>
      <td>√</td>
    </tr>
    <tr>
      <td>timm_models</td>
      <td>res2net50_14w_8s</td>
      <td>single</td>
      <td>X</td>
      <td>√</td>
    </tr>
    <tr>
      <td>timm_models</td>
      <td>res2next50</td>
      <td>single</td>
      <td>X</td>
      <td>√</td>
    </tr>
    <tr>
      <td>timm_models</td>
      <td>resnest101e</td>
      <td>single</td>
      <td>X</td>
      <td>√</td>
    </tr>
    <tr>
      <td>timm_models</td>
      <td>rexnet_100</td>
      <td>single</td>
      <td>X</td>
      <td>√</td>
    </tr>
    <tr>
      <td>timm_models</td>
      <td>selecsls42b</td>
      <td>single</td>
      <td>X</td>
      <td>√</td>
    </tr>
    <tr>
      <td>timm_models</td>
      <td>swsl_resnext101_32x16d</td>
      <td>single</td>
      <td>X</td>
      <td>√</td>
    </tr>
    <tr>
      <td>timm_models</td>
      <td>visformer_small</td>
      <td>single</td>
      <td>X</td>
      <td>√</td>
    </tr>
    <tr>
      <td>timm_models</td>
      <td>volo_d1_224</td>
      <td>single</td>
      <td>X</td>
      <td>√</td>
    </tr>
    <tr>
      <td>timm_models</td>
      <td>xcit_large_24_p8_224</td>
      <td>single</td>
      <td>X</td>
      <td>√</td>
    </tr>
  </tbody>

</table><p>SW info</p><table border="1" class="dataframe table">
  <thead>
    <tr style="text-align: right;">
      <th>name</th>
      <th>target_branch</th>
      <th>target_commit</th>
      <th>refer_branch</th>
      <th>refer_commit</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>torchbench</td>
      <td>main</td>
      <td>d6015d42</td>
      <td>main</td>
      <td>d6015d42</td>
    </tr>
    <tr>
      <td>torch</td>
      <td>main</td>
      <td>88a71594933b2464d9d8b6b3533c5a945a4ac2ff</td>
      <td>main</td>
      <td>bb04f3f66a5b92f0bed3712689f57774f00db349</td>
    </tr>
    <tr>
      <td>torchvision</td>
      <td>main</td>
      <td>0.19.0a0+2c4665f</td>
      <td>main</td>
      <td>0.19.0a0+2c4665f</td>
    </tr>
    <tr>
      <td>torchtext</td>
      <td>main</td>
      <td>0.16.0a0+b0ebddc</td>
      <td>main</td>
      <td>0.16.0a0+b0ebddc</td>
    </tr>
    <tr>
      <td>torchaudio</td>
      <td>main</td>
      <td>2.2.0a0+ea437b3</td>
      <td>main</td>
      <td>2.2.0a0+ea437b3</td>
    </tr>
    <tr>
      <td>torchdata</td>
      <td>main</td>
      <td>0.7.1a0+0790338</td>
      <td>main</td>
      <td>0.7.1a0+0790338</td>
    </tr>
    <tr>
      <td>dynamo_benchmarks</td>
      <td>main</td>
      <td>nightly</td>
      <td>main</td>
      <td>nightly</td>
    </tr>
  </tbody>

</table>

Repro:
[inductor_single_run.sh](https://github.com/chuanqi129/inductor-tools/blob/main/scripts/modelbench/inductor_single_run.sh)
bash inductor_single_run.sh **thread** inference accuracy **suite** **model** float32/amp first static/dynamic
Suspected guilty commit: https://github.com/pytorch/pytorch/commit/7a7853446835b2e88c7e7618d18e9e15ca1029ce
[torchbench-dcgan-inference-float32-static-cpp-multiple-accuracy-crash_guilty_commit.log](https://github.com/pytorch/pytorch/files/15012717/torchbench-dcgan-inference-float32-static-cpp-multiple-accuracy-crash_guilty_commit.log)
cc @ezyang @gchanan @zou3519 @kadeng @msaroufim @bdhirsh @anijain2305 @chauhang @WeizhuoZhang-intel @chuanqi129

**Comment by chuanqi129** [comment]
I'm wonder why inductor CPU FP32 accuracy tests were not gating this regression in CI test for those multiple thread failures. We need to figure out it @leslie-fang-intel @zxd1997066

**Comment by leslie-fang-intel** [comment]
> I'm wonder why inductor CPU FP32 accuracy tests were not gating this regression in CI test for those multiple thread failures. We need to figure out it @leslie-fang-intel @zxd1997066

Since @chuanqi129 is taking annual leave, @zxd1997066 could you help to follow up figuring out why the preCI didn't gate these failures? The hud https://hud.pytorch.org/pr/123399 is all green to land the PR which seems causing the regression.

**Comment by zxd1997066** [comment]
I verified rexnet_100 manually, the guilty commit is right. And the model does pass in CI https://github.com/pytorch/pytorch/actions/runs/8609192929/job/23593107462#step:19:1210. How can I further check it?

**Comment by leslie-fang-intel** [comment]
Maybe to compare the env difference between preCI and your manual testing environment?

**Comment by leslie-fang-intel** [comment]
Hi @yf225, could you take a look of this issue? Sorry that the preCI didn't gate the failures.

**Comment by leslie-fang-intel** [comment]
Hi @yf225, looking into this issue with RN50 FP32 as example. 
```
bash inductor_single_run.sh multiple inference accuracy torchbench resnet50 float32 first static
```

Here are some of my finding:

- Before this commit, the generated FX graph by dynamo is: https://gist.github.com/leslie-fang-intel/7296d66caea0413d492556487c9b58eb#file-success_generated_module-py
- After this commit, the generated FX graph by dynamo is:  https://gist.github.com/leslie-fang-intel/7296d66caea0413d492556487c9b58eb#file-error_generated_module-py

As you can see, the graph has duplicated one copy of the FX nodes entirely. In the Inductor process, when we perform [conv-bn folding](https://github.com/pytorch/pytorch/blob/3af12447f85dfede191a113c052e58fa7b21a8b3/torch/_inductor/fx_passes/pre_grad.py#L296), this results in folding the batch normalization (BN) layer twice for a single convolutional (conv) node, leading to accuracy failures.

**Comment by leslie-fang-intel** [comment]
Since there are several models failed in different scenarios among FP32/AMP Static/Dynamic shape, mark it as high priority. cc @jgong5 @Guobing-Chen @malfet

**Comment by leslie-fang-intel** [comment]
Fixed in https://github.com/pytorch/pytorch/pull/124808

**Comment by leslie-fang-intel** [comment]
@zxd1997066, please help to verify. I think it should be fixed by this week's testing.

**Comment by zxd1997066** [comment]
yes, the issue is fixed in latest report.
